File: DSA_Lab_3/Ex_1.cpp
#include <iostream>
using namespace std;

#define size 25

class Stack {
    int *array;
    int ptr;

public:
    Stack() {
        array = new int[size];
        ptr = -1;
    }
    void push(int x) {
        if (isFull()) {
            cout << "Stack is Full. Can't push any element" << endl;
            return;
        }
        ptr++;
        array[ptr] = x;
    }
    void pop() {
        if (isEmpty()) {
            cout << "Stack is empty. Can't POP" << endl;
            return;
        }
        ptr--;
        cout << "POP done successfuly" << endl;
    }
    int top() {
        if (!isEmpty()) {
            return array[ptr];
        }
        return -1;
    }
    bool isFull() {
        return (ptr == size-1);
    }
    bool isEmpty() {
        return (ptr == -1);
    }
    void crrSit() {
        cout << "----------------------------------" << endl;
        cout << "Total size of Stack = " << size << endl;
        cout << "Current elements in Stack = " << ptr+1 << endl;
        cout << "Available space in Stack = " << size - (ptr+1) << endl;
        cout << "----------------------------------" << endl;
    }
    ~Stack() {
        delete[] array;
    }
};

void menu() {
    cout << "----------------------------------" << endl;
    cout << "Press 1 to push" << endl;
    cout << "Press 2 to pop" << endl;
    cout << "Press 3 to check the top most element in stack" << endl;
    cout << "Press 4 to check that wheather Stack is full or not" << endl;
    cout << "Press 5 to check that wheather Stack is empty or not" << endl;
    cout << "Press 6 to check the current condition of stack" << endl;
    cout << "Press 9 to quit the program" << endl;
    cout << "----------------------------------" << endl;
    cout << "Enter your choice: ";
}

int main() {
    Stack *st = new Stack();
    int choice, element;

    do {
        menu();
        cin >> choice;

        if (choice == 9) {
            break;
        }

        switch(choice) {
            case 1:
                cout << "Enter the element: ";
                cin >> element;
                st->push(element);
                break;
            case 2:
                st->pop();
                break;
            case 3:
                if (st->top() != -1) {
                    cout << "Top most element is " << st->top() << endl;
                } else {
                    cout << "Stack is Empty" << endl;
                }
                break;
            case 4:
                if (st->isFull()) {
                    cout << "Stack is Full" << endl;
                } else {
                    cout << "Stack is not Full" << endl;
                }
                break;
            case 5:
                if (st->isEmpty()) {
                    cout << "Stack is empty" << endl;
                } else {
                    cout << "Stack is not empty" << endl;
                }
                break;
            case 6:
                st->crrSit();
                break;
            default:
                cout << "Invalid option selected" << endl;

        }
    } while(choice != 9);

    delete st;
    st = nullptr;
    system("pause");
    return 0;
}

File: DSA_Lab_3/output/Mustansar - 1071/EX_1_Using GUI (JAVA)/Source Code/EX_1.java
import javax.swing.JOptionPane;
class Stack {
    private int ptr;
    private int arraySize;
    private int[] arr;

    public Stack(int size) {
        this.arraySize = size;
        this.arr = new int[arraySize];
        this.ptr = -1;
    }
    public void push(int x) {
        if (isFull()) {
            JOptionPane.showMessageDialog(null, "Stack Overflow", "Stack", JOptionPane.ERROR_MESSAGE);
            return;
        }
        this.ptr++;
        arr[ptr] = x;
        JOptionPane.showMessageDialog(null, "Element pushed successfully", "Stack", JOptionPane.INFORMATION_MESSAGE);
    }
    public void pop() {
        if (isEmpty()) {
            JOptionPane.showMessageDialog(null, "Stack is Empty so cannot pop", "Stack", JOptionPane.ERROR_MESSAGE);
            return;
        }
        this.ptr--;
        JOptionPane.showMessageDialog(null, "Element Popped successfully", "Stack", JOptionPane.INFORMATION_MESSAGE);
    }
    public void top() {
        if (isEmpty()) {
            JOptionPane.showMessageDialog(null, "Stack is Empty so no element is at Top", "Stack", JOptionPane.ERROR_MESSAGE);
            return;
        }
        JOptionPane.showMessageDialog(null, "Top element is " + this.arr[ptr], "Stack", JOptionPane.INFORMATION_MESSAGE);
    }
    public boolean isEmpty() {
        return (this.ptr == -1);
    }
    public boolean isFull() {
        return (this.ptr == this.arraySize-1);
    }
    public void currentStatus() {
        String status = "Total space in stack = " + this.arraySize
                + "\nUsed space in stack = " + (this.ptr+1)
                + "\nAvailable space in stack = " + (this.arraySize - (this.ptr+1));
        JOptionPane.showMessageDialog(null, status, "Stack", JOptionPane.INFORMATION_MESSAGE);
    }
}
public class EX_1 {
    public static void main(String[] args) {
        String choiceStr, elementStr;
        int choice, element;
        Stack st = new Stack(25);
        while (true){
            choiceStr = JOptionPane.showInputDialog(menu());
            if (choiceStr == null) {
                JOptionPane.showMessageDialog(null, "Exiting the programme", "Stack", JOptionPane.INFORMATION_MESSAGE);
                System.exit(0);
            }
            choice = 0;
            try {
                choice = Integer.parseInt(choiceStr);
            } catch (NumberFormatException e) {
                JOptionPane.showMessageDialog(null, "Invalid Input: Please enter a valid integer choice", "Stack", JOptionPane.ERROR_MESSAGE);
            }


            if (choice == 9) {
                JOptionPane.showMessageDialog(null, "Exiting the programme", "Stack", JOptionPane.INFORMATION_MESSAGE);
                break;
            }

            if (!(choice >= 1 && choice <= 6)) {
                JOptionPane.showMessageDialog(null, "Invalid Choice", "Stack", JOptionPane.ERROR_MESSAGE);
            }

            switch (choice) {
                case 1:
                    elementStr = JOptionPane.showInputDialog("Enter the number: ");
                    element = Integer.parseInt(elementStr);
                    st.push(element);
                    break;
                case 2:
                    st.pop();
                    break;
                case 3:
                    st.top();
                    break;
                case 4:
                    if (st.isEmpty()) {
                        JOptionPane.showMessageDialog(null, "Stack is Empty", "Stack", JOptionPane.INFORMATION_MESSAGE);
                    } else {
                        JOptionPane.showMessageDialog(null, "Stack is not Empty", "Stack", JOptionPane.INFORMATION_MESSAGE);
                    }
                    break;
                case 5:
                    if (st.isFull()) {
                        JOptionPane.showMessageDialog(null, "Stack is Full", "Stack", JOptionPane.INFORMATION_MESSAGE);
                    }
                    else {
                        JOptionPane.showMessageDialog(null, "Stack is not Full", "Stack", JOptionPane.INFORMATION_MESSAGE);
                    }
                    break;
                case 6:
                    st.currentStatus();
                    break;
            }
        }
    }
    public static String menu() {
        return ("""
                Press 1 to Push Element in Stack
                Press 2 to Pop the Element from Stack
                Press 3 to Check the Top Element in the Stack
                Press 4 to Check whether Stack is Empty
                Press 5 to Check whether Stack is Full
                Press 6 to check current status of Stack
                Press 9 to Quit the programme""");
    }
}